lightning/src/ln/peers/handshake/states.rs

```diff
@@ -1,30 +1,881 @@
-	pub(super) hash: HandshakeHash,
-	pub(super) chaining_key: [u8; 32],
-	pub(super) temporary_key: [u8; 32],
-	pub(super) ephemeral_private_key: SecretKey,
+// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#handshake-state-initialization
+fn handshake_state_initialization(responder_static_public_key: &PublicKey) -> (Sha256, Sha256) {
+	let protocol_name = b"Noise_XK_secp256k1_ChaChaPoly_SHA256";
+	let prologue = b"lightning";
+
+	// 1. h = SHA-256(protocolName)
+	// 2. ck = h
+	let chaining_key = concat_then_sha256!(protocol_name);
+
+	// 3. h = SHA-256(h || prologue)
+	let hash = concat_then_sha256!(chaining_key, prologue);
+
+	// h = SHA-256(h || responderPublicKey)
+	let hash = concat_then_sha256!(hash, responder_static_public_key.serialize());
+
+	(hash, chaining_key)
 }
 
-pub struct ActThreeExpectation {
-	pub(super) hash: HandshakeHash,
-	pub(super) chaining_key: [u8; 32],
-	pub(super) temporary_key: [u8; 32],
-	pub(super) ephemeral_private_key: SecretKey,
-	pub(super) remote_ephemeral_public_key: PublicKey,
+// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-one (sender)
+// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-two (sender)
+fn calculate_act_message(local_private_ephemeral_key: &SecretKey, local_public_ephemeral_key: &PublicKey, remote_public_key: &PublicKey, chaining_key: ChainingKey, hash: Sha256, act_out: &mut [u8]) -> (Sha256, SymmetricKey, SymmetricKey) {
+	// 1. e = generateKey() (passed in)
+	// 2. h = SHA-256(h || e.pub.serializeCompressed())
+	let serialized_local_public_key = local_public_ephemeral_key.serialize();
+	let hash = concat_then_sha256!(hash, serialized_local_public_key);
+
+	// 3. ACT1: es = ECDH(e.priv, rs)
+	// 3. ACT2: es = ECDH(e.priv, re)
+	let ecdh = ecdh(local_private_ephemeral_key, &remote_public_key);
+
+	// 4. ACT1: ck, temp_k1 = HKDF(ck, es)
+	// 4. ACT2: ck, temp_k2 = HKDF(ck, ee)
+	let (chaining_key, temporary_key) = hkdf::derive(&chaining_key, &ecdh);
+
+	// 5. ACT1: c = encryptWithAD(temp_k1, 0, h, zero)
+	// 5. ACT2: c = encryptWithAD(temp_k2, 0, h, zero)
+	chacha::encrypt(&temporary_key, 0, &hash, &[0; 0], &mut act_out[34..]);
+
+	// 6. h = SHA-256(h || c)
+	let hash = concat_then_sha256!(hash, &act_out[34..]);
+
+	// Send m = 0 || e.pub.serializeCompressed() || c
+	act_out[1..34].copy_from_slice(&serialized_local_public_key);
+
+	(hash, chaining_key, temporary_key)
 }
+
+// Due to the very high similarity of acts 1 and 2, this method is used to process both
+// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-one (receiver)
+// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-two (receiver)
+fn process_act_message(act_bytes: &[u8], local_private_key: &SecretKey, chaining_key: ChainingKey, hash: Sha256) -> Result<(PublicKey, Sha256, SymmetricKey, SymmetricKey), String> {
+	// 1. Read exactly 50 bytes from the network buffer
```

Comment:
You can `enforced in state translation respective act processing method` or anything else to indicate where the check is actually enforced.

Reply:
Great feedback. It isn't great that a mismatch would fail with a slice error instead of something a bit more clear. I've added asserts and a comment here to make it more clear the relationship between this function and partial act messages.

Resolved in `review: Non-controversial renames, nits, small fixups`